### PR TITLE
Updating FD boot loader for 2.0

### PR DESCRIPTION
**Pull request by abhullar-tt**

### CODE REVIEW

File: tt_metal/impl/dispatch/command_queue_interface.hpp [hunk 1]
```diff
@@ -114,14 +137,31 @@ struct SystemMemoryCQInterface {
...
+
+            25% of this = 65,535
+
+            compleion queue size = 65,535 * 4096 = 268,431,360
+
+            issue queue size = 805,310,400
+
+      */
+
+      command_completion_region_size(268431360),
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/command_queue_interface.hpp:156]: hardcoded sizes need to be updated

File: tt_metal/hostdevcommon/common_runtime_address_map.h [hunk 2]
```diff
@@ -123,51 +123,24 @@ constexpr static std::uint32_t DISPATCH_MESSAGE_ADDR = RING_BUFFER_ADDR + RING_B
...
 constexpr static std::uint64_t DISPATCH_MESSAGE_REMOTE_SENDER_ADDR = DISPATCH_MESSAGE_ADDR + DISPATCH_MESSAGE_ADDR_RESERVATION;
 
-constexpr static std::uint32_t COMMAND_PTR_SHARD_IDX = 8;
-
 // Command queue pointers
-constexpr static uint32_t CQ_ISSUE_READ_PTR = DISPATCH_MESSAGE_REMOTE_SENDER_ADDR + DISPATCH_MESSAGE_ADDR_RESERVATION;
-constexpr static uint32_t CQ_ADDR_RESERVATION = 96;
-constexpr static uint32_t CQ_ISSUE_WRITE_PTR = CQ_ISSUE_READ_PTR + L1_ALIGNMENT;
-constexpr static uint32_t CQ_COMPLETION_WRITE_PTR = CQ_ISSUE_WRITE_PTR + L1_ALIGNMENT;
+constexpr static uint32_t CQ_PREFETCH_Q_RD_PTR = DISPATCH_MESSAGE_REMOTE_SENDER_ADDR + DISPATCH_MESSAGE_ADDR_RESERVATION;
```
Comment (abhullar-tt) [inline tt_metal/hostdevcommon/common_runtime_address_map.h:127]: move all dispatch related addresses to dispatch code

File: tt_metal/impl/device/device.cpp [hunk 3]
```diff
@@ -319,22 +319,174 @@ void Device::compile_command_queue_programs() {
...
+
+    uint32_t prefetch_d_buffer_pages = PREFETCH_D_BUFFER_SIZE >> PREFETCH_D_BUFFER_LOG_PAGE_SIZE;
+
+    constexpr uint32_t prefetch_sync_sem = 0;
+    constexpr uint32_t prefetch_downstream_cb_sem = 1;
+    constexpr uint32_t dispatch_sync_sem = 0;
+    constexpr uint32_t dispatch_cb_sem = 1;
+
+    constexpr uint32_t prefetch_d_upstream_cb_sem = 1;
+    constexpr uint32_t prefetch_d_downstream_cb_sem = 2;
```
Comment (abhullar-tt) [inline tt_metal/impl/device/device.cpp:343]: these are sem ids

File: tt_metal/impl/dispatch/dispatch_core_manager.hpp [hunk 4]
```diff
@@ -12,20 +12,20 @@ namespace tt::tt_metal {
...
+    std::optional<tt_cxy_pair> prefetcher = std::nullopt;  // Pulls commands from the issue queue for a given command queue on a device
     std::optional<tt_cxy_pair> completion_queue_writer = std::nullopt; // Pushes to completion queue for a given command queue on a device
-    std::optional<tt_cxy_pair> command_dispatcher = std::nullopt; // Relays work to worker cores on device that command is targeting. Currently for MMIO devices, command_dispatcher == completion_queue_writer
-    std::optional<tt_cxy_pair> remote_push_and_pull = std::nullopt;    // Receives fast dispatch commands from DST ethernet router, relays programs to command_dispatcher, and returns data through SRC router
+    std::optional<tt_cxy_pair> dispatcher = std::nullopt; // Relays work to worker cores on device that command is targeting. Currently for MMIO devices, dispatcher == completion_queue_writer
+    std::optional<tt_cxy_pair> remote_push_and_pull = std::nullopt;    // Receives fast dispatch commands from DST ethernet router, relays programs to dispatcher, and returns data through SRC router
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/dispatch_core_manager.hpp:28]: legacy from 1.x

File: tt_metal/impl/dispatch/command_queue_interface.hpp [hunk 5]
```diff
@@ -425,8 +437,44 @@ class SystemMemoryManager {
...
+            while (this->prefetch_q_dev_ptrs[cq_id] == this->prefetch_q_dev_fences[cq_id]) {
+                tt::Cluster::instance().read_core(&fence, sizeof(uint32_t), this->prefetcher_cores[cq_id], CQ_PREFETCH_Q_RD_PTR);
+                this->prefetch_q_dev_fences[cq_id] = fence;
+            }
+
+            // std::cout << "got space again in fetchq" << std::endl;
+        }
     }
 
+    void fetch_queue_push_back(uint16_t command_size_16B, const uint8_t cq_id) {
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/command_queue_interface.hpp:467]: consider a rename because we don't allocate any new space in fetch queue

Comment (abhullar-tt) [reply tt_metal/impl/dispatch/command_queue_interface.hpp:467]: fetch_queue_write (similar to cq_write)

**Comment by aliuTT** [comment]
@abhullar-tt you deleted all the command queue files? Are we going to port Grayskull dispatch to FD2 as well?

File: tt_metal/impl/device/device.cpp [hunk 6]
```diff
@@ -319,22 +319,174 @@ void Device::compile_command_queue_programs() {
...
+                tt_cxy_pair prefetcher_location = dispatch_core_manager::get(num_hw_cqs).prefetcher_core(device_id, channel, cq_id);
+                tt_cxy_pair completion_q_writer_location = dispatch_core_manager::get(num_hw_cqs).completion_queue_writer_core(device_id, channel, cq_id);
+                tt_cxy_pair dispatch_location = dispatch_core_manager::get(num_hw_cqs).dispatcher_core(device_id, channel, cq_id);
+
+                TT_ASSERT(prefetcher_location.chip == this->id() and completion_q_writer_location.chip == this->id(),
+                    "Issue queue interface is on device {} and completion queue interface is on device {} but they are expected to be on device {}", prefetcher_location.chip, completion_q_writer_location.chip, this->id());
+
+                CoreCoord prefetcher_physical_core = get_physical_core_coordinate(prefetcher_location, dispatch_core_type);
+                CoreCoord completion_q_physical_core = get_physical_core_coordinate(completion_q_writer_location, dispatch_core_type);
+                CoreCoord dispatch_physical_core = get_physical_core_coordinate(dispatch_location, dispatch_core_type);
```
Comment (pgkeller) [inline tt_metal/impl/device/device.cpp:367]: there's an inconsistency between "prefetcher" and "dispatch"

**Comment by abhullar-tt** [comment]
> @abhullar-tt you deleted all the command queue files? Are we going to port Grayskull dispatch to FD2 as well?

I don't think we mean to keep the grayskull fork around

**Comment by abhullar-tt** [comment]
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8517169399

**Pull request merged**